File: app/src/main/java/com/mindorks/bootcamp/learndagger/di/modules/ApplicationModule.java
package com.mindorks.bootcamp.learndagger.di.modules;

import android.content.Context;

import com.mindorks.bootcamp.learndagger.MyApplication;
import com.mindorks.bootcamp.learndagger.data.local.DatabaseService;
import com.mindorks.bootcamp.learndagger.data.remote.NetworkService;
import com.mindorks.bootcamp.learndagger.di.qualifier.ApiKey;
import com.mindorks.bootcamp.learndagger.di.qualifier.DatabaseName;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class ApplicationModule {

    private MyApplication myApplication ;

    public ApplicationModule(MyApplication myApplication) {
        this.myApplication = myApplication;
    }

    @Provides
    public Context provideContext(){
        return myApplication;
    }

    @Provides
    @DatabaseName
    String provideDatabaseName(){
        return "aaa";
    }
    @Provides
    int provideDatabaseVersion(){
        return  1 ;
    }

    @Provides
    @ApiKey
    String provideApiKey(){
        return "ssss";
    }

/*    @Singleton
    @Provides
    public NetworkService provideNetworkService(){
        return new NetworkService(myApplication , "abc");
    }


    @Singleton
    @Provides
    public DatabaseService provideDatabaseService(){
        return new DatabaseService(myApplication , "zzz" , 1);
    }*/


}

File: app/src/main/java/com/mindorks/bootcamp/learndagger/di/modules/ApplicationModule.kt
package com.mindorks.bootcamp.learndagger.di.modules

import android.content.Context
import androidx.room.Room
import com.mindorks.bootcamp.demo.di.ApplicationContext
import com.mindorks.bootcamp.demo.di.DatabaseInfo
import com.mindorks.bootcamp.demo.di.NetworkInfo
import com.mindorks.bootcamp.learndagger.BuildConfig
import com.mindorks.bootcamp.learndagger.MyApplication
import com.mindorks.bootcamp.learndagger.data.Networking
import com.mindorks.bootcamp.learndagger.data.local.DatabaseService
import com.mindorks.bootcamp.learndagger.data.local.MIGRATION_1_2
import com.mindorks.bootcamp.learndagger.data.remote.NetworkService
import dagger.Module
import dagger.Provides
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Singleton

@Module
class ApplicationModule (var myApplication: MyApplication){



    @Provides
    @ApplicationContext
    fun provideContext(): Context = myApplication

    @Provides
    @DatabaseInfo
    fun provideDatabaseName(): String {
        return "aaa"
    }

    @Provides
    fun provideDatabaseVersion(): Int {
        return 1
    }

   /* @Provides
    @NetworkInfo
    fun provideApiKey(): String {
        return "ssss"
    }*/

    @Singleton
    @Provides
    fun provideDatabaseService(): DatabaseService = Room.databaseBuilder(
            myApplication,
            DatabaseService::class.java,
            "bootcamp=database-project-db"
    ).addMigrations(MIGRATION_1_2)
            .build()


    @Provides
    fun provideCompositeDisposable():CompositeDisposable = CompositeDisposable()


    @Singleton
    @Provides
    fun provideNetworkService(): NetworkService = Networking.create(
            BuildConfig.API_KEY,
            BuildConfig.BASE_URL,
            myApplication.cacheDir,
            10 * 1024 * 1024
    )
}